crates/driver/src/domain/competition/sorting.rs

```diff
@@ -0,0 +1,95 @@
+use {
+    crate::{
+        domain::{
+            competition::{auction::Tokens, order},
+            eth,
+        },
+        util,
+    },
+    chrono::{Duration, Utc},
+    std::sync::Arc,
+};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub enum SortingKey {
```

Comment:
Does this need to be public?

It's not possible in the code, but how would this behave if one strategy returned `Int` and `Bool` for different items? Shouldn't each strategy be forced by the type-system to return a common type?

Reply:
> if one strategy returned Int and Bool for different items?

I didn't quite get the problem.

> Shouldn't each strategy be forced by the type-system to return a common type?

Initially, I tried to use `impl/dyn Ord` trait, but it is not possible since it requires `Self`, so I had to define a common enum.

Reply:
What would happen if someone implements

```
impl SortingStrategy for OrderClass {
    fn key(&self, order: &order::Order, _tokens: &Tokens, _solver: &eth::H160) -> SortingKey {
        match order.kind {
            order::Kind::Market => SortingKey::Int(2),
            _ => SortingKey::Bool(true),
```

This would be allowed by the type-system but I'm not even sure how the sorting would work in this case.

I was hoping we can get the trait enforce that each implementation of it returns a single unique type, but I'm not sure this is possible (maybe @Mateo-mro or @MartinquaXD who know rust much better than me)

Reply:
Good question. I experimented a bit with enforcing this in the type system but I was not able to make it work.
Things I tried:
* return an associated type from the trait
* return `impl Into<SortingKey>` from the trait
* use generic parameter on the trait

But even if we could enforce that one function always returns the same enum variant there is technically still the error case where for one order we apply strategy 1 first and for another order strategy 2 which would still in 2 different enum variants getting compared to each other.
I assume enforcing this in the type system has to be possible somehow but I think we'd have to jump through unreasonably many hoops.

I think the pattern @squadgazzz used here is probably the best bang for the buck in terms of type safety:
```rust
pub struct OrderClass;
impl SortingStrategy for OrderClass {
    fn key(&self, order: &order::Order, _tokens: &Tokens, _solver: &eth::H160) -> SortingKey {
        SortingKey::Int(match order.kind {
            order::Kind::Market => 2,
            order::Kind::Limit { .. } => 1,
            order::Kind::Liquidity => 0,
        })
    }
}
```
Note that the match statement is inside the enum variant so that enforces that the correct type has to be provided.
If this pattern gets applied everywhere (`Bool` being the only exception at the moment) we should be okay.

Reply:
We could also have a runtime check that enforces correct comparisons by manually implementing `Ord` and `PartialOrd`:
```rust
impl Ord for SortingKey {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match (self, other) {
            (Self::Int(a), Self::Int(b)) => a.cmp(b),
            (Self::BigRational(a), Self::BigRational(b)) => a.cmp(b),
            (Self::Timestamp(a), Self::Timestamp(b)) => a.cmp(b),
            (Self::Bool(a), Self::Bool(b)) => a.cmp(b),
            _ => panic!("can't compare different types, dummy")
        }
    }
}

impl PartialOrd<Self> for SortingKey {
   fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
       Some(self.cmp(other))
   }
}
```